### PR TITLE
AI: extract pathfinding special actions and rules to separate files

### PR DESCRIPTION
**Pull request by nullkiller**
Small refactoring.

### CODE REVIEW

File: AI/VCAI/Pathfinding/Actions/BoatActions.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+* License: GNU General Public License v2.0 or later
+* Full text of license available in license.txt file, in main folder
+*
+*/
+
+#include "StdInc.h"
+#include "../../Goals/AdventureSpellCast.h"
+#include "../../Goals/BuildBoat.h"
+#include "../../../../lib/mapping/CMap.h"
+#include "../../../../lib/mapObjects/MapObjects.h"
```
Comment (nullkiller) [inline AI/VCAI/Pathfinding/Actions/BoatActions.cpp]: Can i reference it in some simplier way? Like #include "mapping/CMap.h". It should be already in some include path

Comment (alexvins) [reply AI/VCAI/Pathfinding/Actions/BoatActions.cpp]: For now it should stay relative path.

File: AI/VCAI/Pathfinding/Actions/ISpecialAction.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+{
+public:
+	virtual Goals::TSubgoal whatToDo(HeroPtr hero) const = 0;
+
+	virtual void applyOnDestination(
+		HeroPtr hero,
+		CDestinationNodeInfo & destination,
+		const PathNodeInfo & source,
+		AIPathNode * dstMode,
+		const AIPathNode * srcNode) const
```
Comment (alexvins) [inline AI/VCAI/Pathfinding/Actions/ISpecialAction.h]: Missing forward declaration of `AIPathNode`

**Comment by alexvins** [comment]
Please rebase on develop once more, it should fix build.

File: AI/VCAI/Pathfinding/Rules/AILayerTransitionRule.cpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+		if(hero->canCastThisSpell(summonBoatSpell)
+			&& hero->getSpellSchoolLevel(summonBoatSpell) >= SecSkillLevel::ADVANCED)
+		{
+			// TODO: For lower school level we might need to check the existance of some boat
+			summonableVirtualBoat.reset(new SummonBoatAction());
+		}
+	}
+
+	std::shared_ptr<const VirtualBoatAction> AILayerTransitionRule::findVirtualBoat(
+		CDestinationNodeInfo &destination,
```
Comment (alexvins) [inline AI/VCAI/Pathfinding/Rules/AILayerTransitionRule.cpp]: Put space around `&`

File: AI/VCAI/Pathfinding/Rules/AILayerTransitionRule.cpp [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+			&& summonableVirtualBoat->isAffordableBy(nodeStorage->getHero(), nodeStorage->getAINode(source.node)))
+		{
+			virtualBoat = summonableVirtualBoat;
+		}
+
+		return virtualBoat;
+	}
+
+	bool AILayerTransitionRule::tryEmbarkVirtualBoat(
+		CDestinationNodeInfo &destination,
```
Comment (alexvins) [inline AI/VCAI/Pathfinding/Rules/AILayerTransitionRule.cpp]: Put space around `&`

File: AI/VCAI/Pathfinding/Rules/AILayerTransitionRule.h [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+			const PathNodeInfo & source,
+			CDestinationNodeInfo & destination,
+			const PathfinderConfig * pathfinderConfig,
+			CPathfinderHelper * pathfinderHelper) const override;
+
+	private:
+		void setup();
+
+		std::shared_ptr<const VirtualBoatAction> findVirtualBoat(
+			CDestinationNodeInfo &destination,
```
Comment (alexvins) [inline AI/VCAI/Pathfinding/Rules/AILayerTransitionRule.h]: Put space around `&`

File: AI/VCAI/Pathfinding/Actions/BoatActions.cpp [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+#include "BoatActions.h"
+
+namespace AIPathfinding
+{
+	Goals::TSubgoal BuildBoatAction::whatToDo(HeroPtr hero) const
```
Comment (alexvins) [inline AI/VCAI/Pathfinding/Actions/BoatActions.cpp]: Copying `HeroPtr` is not that cheap. As long as all of this is changed, I`d like to see more usage of constant references.

Comment (nullkiller) [reply AI/VCAI/Pathfinding/Actions/BoatActions.cpp]: Fixed

File: AI/VCAI/Pathfinding/Rules/AILayerTransitionRule.h (hunk 5 above)
Comment (nullkiller) [reply AI/VCAI/Pathfinding/Rules/AILayerTransitionRule.h]: Fixed

File: AI/VCAI/Pathfinding/Rules/AILayerTransitionRule.cpp (hunk 4 above)
Comment (nullkiller) [reply AI/VCAI/Pathfinding/Rules/AILayerTransitionRule.cpp]: Fixed

File: AI/VCAI/Pathfinding/Rules/AILayerTransitionRule.cpp (hunk 3 above)
Comment (nullkiller) [reply AI/VCAI/Pathfinding/Rules/AILayerTransitionRule.cpp]: Fixed

**Pull request merged**